tsdb/chunks/chunk_write_queue.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunks
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+)
+
+type chunkWriteJob struct {
+	cutFile   bool
+	seriesRef HeadSeriesRef
+	mint      int64
+	maxt      int64
+	chk       chunkenc.Chunk
+	ref       ChunkDiskMapperRef
+	callback  func(error)
+}
+
+var (
+	queueOperationAdd      = "add"
+	queueOperationGet      = "get"
+	queueOperationComplete = "complete"
+	queueOperations        = []string{queueOperationAdd, queueOperationGet, queueOperationComplete}
+)
+
+// chunkWriteQueue is a queue for writing chunks to disk in a non-blocking fashion.
+// Chunks that shall be written get added to the queue, which is consumed asynchronously.
+// Adding jobs to the job is non-blocking as long as the queue isn't full.
+type chunkWriteQueue struct {
+	size int
+	jobs chan chunkWriteJob
+
+	chunkRefMapMtx sync.RWMutex
+	chunkRefMap    map[ChunkDiskMapperRef]chunkenc.Chunk
+
+	isRunningMtx sync.Mutex // protects the isRunning property.
+	isRunning    bool       // used to prevent that new jobs get added to the queue when the chan is already closed.
+
+	workerWg sync.WaitGroup
+
+	writeChunk writeChunkF
+
+	operationsMetric *prometheus.CounterVec
+}
+
+// writeChunkF is a function which writes chunks, it is dynamic to allow mocking in tests.
+type writeChunkF func(HeadSeriesRef, int64, int64, chunkenc.Chunk, ChunkDiskMapperRef, bool) error
+
+func newChunkWriteQueue(reg prometheus.Registerer, size int, writeChunk writeChunkF) *chunkWriteQueue {
+	q := &chunkWriteQueue{
+		size:        size,
+		jobs:        make(chan chunkWriteJob, size),
+		chunkRefMap: make(map[ChunkDiskMapperRef]chunkenc.Chunk, size),
+		writeChunk:  writeChunk,
+
+		operationsMetric: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "prometheus_tsdb_chunk_write_queue_operations_total",
+				Help: "Number of operations on the chunk_write_queue.",
+			},
+			[]string{"operation"},
+		),
+	}
+
+	if reg != nil {
+		reg.MustRegister(q.operationsMetric)
+
+		// Initialize series for all the possible labels.
+		for _, op := range queueOperations {
+			q.operationsMetric.WithLabelValues(op)
+		}
```

Comment:
Since you're optimising a hot path here with potentially thousands of concurrent calls, and these labels are known beforehand, I think it's worth creating separate fields for each one of the counters in the queue itself, like:

```go
	q.adds = q.operationsMetric.WithLabelValues(queueOperationAdd)
	q.gets = q.operationsMetric.WithLabelValues(queueOperationGet)
	q.completed = q.operationsMetric.WithLabelValues(queueOperationComplete)
```

Because `WithLabelValues` has to acquire a read mutex itself (and lookup in a map), plus labels hashing. There's pretty much difference:
```
name                                                               old time/op  new time/op  delta
ChunkWriteQueue_addJob/with_reads_false/1_concurrent_writes-16      256ns ± 2%   145ns ± 1%  -43.38%  (p=0.008 n=5+5)
ChunkWriteQueue_addJob/with_reads_false/10_concurrent_writes-16     281ns ± 1%   174ns ± 1%  -38.09%  (p=0.008 n=5+5)
ChunkWriteQueue_addJob/with_reads_false/100_concurrent_writes-16    290ns ± 0%   181ns ± 1%  -37.50%  (p=0.008 n=5+5)
ChunkWriteQueue_addJob/with_reads_false/1000_concurrent_writes-16   302ns ± 1%   182ns ± 2%  -39.67%  (p=0.008 n=5+5)
ChunkWriteQueue_addJob/with_reads_true/1_concurrent_writes-16       296ns ± 1%   166ns ± 2%  -43.80%  (p=0.016 n=4+5)
ChunkWriteQueue_addJob/with_reads_true/10_concurrent_writes-16      342ns ± 9%   209ns ± 1%  -39.09%  (p=0.008 n=5+5)
ChunkWriteQueue_addJob/with_reads_true/100_concurrent_writes-16     341ns ± 9%   208ns ± 1%  -38.86%  (p=0.008 n=5+5)
ChunkWriteQueue_addJob/with_reads_true/1000_concurrent_writes-16    361ns ± 8%   223ns ± 8%  -38.12%  (p=0.008 n=5+5)
```

Reply:
That's another great idea. 
Even though in this case the time savings are outside the lock scope I think this still makes sense, because we have seen real world situations where millions of calls to `ChunkDiskMapper.WriteChunk()` happened almost at once and in such a situations this improvement should help even if it only shortens the call to `.WriteChunk()` without shortening the time during which the lock is held.

Done: e0c1ca37a8f8bd64b6dbb3310b16888ff7441f13